File: mobile/src/main/java/com/windscribe/mobile/upgradeactivity/UpgradeActivity.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.upgradeactivity;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static com.windscribe.mobile.email.AddEmailActivity.goToHomeAfterFinish;
import static com.windscribe.vpn.Windscribe.appContext;
import static com.windscribe.vpn.constants.ExtraConstants.PROMO_EXTRA;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.amazon.device.iap.model.Product;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.ProductDetails;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.QueryProductDetailsParams;
import com.google.common.collect.ImmutableList;
import com.windscribe.mobile.R;
import com.windscribe.mobile.base.BaseActivity;
import com.windscribe.mobile.confirmemail.ConfirmActivity;
import com.windscribe.mobile.custom_view.CustomDialog;
import com.windscribe.mobile.custom_view.ErrorFragment;
import com.windscribe.mobile.custom_view.ProgressFragment;
import com.windscribe.mobile.di.ActivityModule;
import com.windscribe.mobile.email.AddEmailActivity;
import com.windscribe.mobile.welcome.WelcomeActivity;
import com.windscribe.mobile.windscribe.WindscribeActivity;
import com.windscribe.vpn.Windscribe;
import com.windscribe.vpn.api.response.PushNotificationAction;
import com.windscribe.vpn.billing.AmazonBillingManager;
import com.windscribe.vpn.billing.BillingFragmentCallback;
import com.windscribe.vpn.billing.GoogleBillingManager;
import com.windscribe.vpn.billing.PurchaseState;
import com.windscribe.vpn.billing.WindscribeInAppProduct;
import com.windscribe.vpn.constants.NetworkKeyConstants;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import javax.inject.Inject;

public class UpgradeActivity extends BaseActivity
        implements UpgradeView
        , BillingFragmentCallback {

    public enum BillingType {
        Google("android"), Amazon("amazon");

        private final String value;

        BillingType(String value) {
            this.value = value;
        }

        @SuppressWarnings("unused")
        public String getValue() {
            return value;
        }
    }

    private static final String TAG = "upgrade_a";

    @Inject
    CustomDialog mUpgradeDialog;

    @Inject
    UpgradePresenter mUpgradePresenter;

    private boolean bBillingProcessFinished = false;

    private BillingType billingType = BillingType.Google;

    private AmazonBillingManager mAmazonBillingManager;

    private GoogleBillingManager mGoogleBillingManager;

    private final Logger mUpgradeLog = LoggerFactory.getLogger(TAG);

    private ProductDetails selectedProductDetails;

    private Boolean upgradingFromWebsite = false;

    public static Intent getStartIntent(Context context) {
        return new Intent(context, UpgradeActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setActivityModule(new ActivityModule(this, this)).inject(this);
        setContentLayout(R.layout.activity_upgrade, false);

        mUpgradeLog.info("OnCreate: Upgrade Activity");
        showProgressBar("Loading Billing Plans...");
        // In app notification promo
        if (getIntent().hasExtra(PROMO_EXTRA)) {
            PushNotificationAction pushNotificationAction = (PushNotificationAction) getIntent()
                    .getSerializableExtra(PROMO_EXTRA);
                mUpgradePresenter.setPushNotificationAction(pushNotificationAction);
        }else{
            // Push notification promo
            if(appContext.appLifeCycleObserver.getPushNotificationAction() != null){
                mUpgradePresenter.setPushNotificationAction(appContext.appLifeCycleObserver.getPushNotificationAction());
            }
        }
        setBillingType();
        if (billingType == BillingType.Amazon) {
            mAmazonBillingManager = Windscribe.getAppContext().getAmazonBillingManager();
            getLifecycle().addObserver(mAmazonBillingManager);
            initAmazonBillingLifecycleListeners();
        } else {
            mGoogleBillingManager = Windscribe.getAppContext().getBillingManager();
            getLifecycle().addObserver(mGoogleBillingManager);
            initBillingLifecycleListeners();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if(upgradingFromWebsite){
            goBackToMainActivity();
        }else{
            if (billingType == BillingType.Google) {
                mUpgradePresenter.checkBillingProcessStatus();
            }
            mUpgradePresenter.setLayoutFromApiSession();
        }
    }

    @Override
    protected void onDestroy() {
        mUpgradeDialog.cancel();
        mUpgradePresenter.onDestroy();
        super.onDestroy();
    }

    @Override
    public BillingType getBillingType() {
        return billingType;
    }

    @Override
    public void getProducts(List<String> skuList) {
        mAmazonBillingManager.getProducts(skuList);
    }

    @Override
    public void goBackToMainActivity() {
        mUpgradeLog.info("Going back to previous activity...");
        onBackPressed();
    }

    @Override
    public void goToAddEmail() {
        Intent startIntent = new Intent(this, AddEmailActivity.class);
        startIntent.putExtra(goToHomeAfterFinish, false);
        startActivity(startIntent);
    }

    @Override
    public void goToConfirmEmail() {
        Intent startIntent = new Intent(this, ConfirmActivity.class);
        startActivity(startIntent);
    }

    @Override
    public void gotToClaimAccount() {
        Intent startIntent = WelcomeActivity.getStartIntent(this);
        startIntent.putExtra("startFragmentName", "AccountSetUp");
        startActivity(startIntent);
        finish();
    }

    @Override
    public void hideProgressBar() {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.cl_upgrade);
        if (fragment instanceof ProgressFragment) {
            getSupportFragmentManager().beginTransaction().remove(fragment).commitNow();
        }
    }

    @Override
    public boolean isBillingProcessFinished() {
        return bBillingProcessFinished;
    }

    @Override
    public void onContinuePlanClick(ProductDetails productDetails, int selectedIndex) {
        if (productDetails != null) {
            mUpgradeLog.info("User clicked on plan item...");
            String regionPlanUrl = mUpgradePresenter.regionalPlanIfAvailable(productDetails.getProductId());
            if (regionPlanUrl != null) {
                mUpgradePresenter.onRegionalPlanSelected(regionPlanUrl);
            } else {
                this.selectedProductDetails = productDetails;
                BillingFlowParams.ProductDetailsParams.Builder builder = BillingFlowParams.ProductDetailsParams.newBuilder();
                builder.setProductDetails(productDetails);
                if (productDetails.getSubscriptionOfferDetails() != null) {
                    String offerToken = productDetails
                            .getSubscriptionOfferDetails()
                            .get(selectedIndex)
                            .getOfferToken();
                    builder.setOfferToken(offerToken);
                }
                ImmutableList<BillingFlowParams.ProductDetailsParams> productDetailsParamsList = ImmutableList.of(builder.build());
                mUpgradePresenter.onMonthlyItemClicked(productDetailsParamsList);
            }
        }
    }

    @Override
    public void openUrlInBrowser(String url) {
        upgradingFromWebsite = true;
        openURLInBrowser(url);
    }

    @Override
    public void onContinuePlanClick(final Product selectedSku) {
        mUpgradePresenter.onContinuePlanClick(selectedSku);
    }

    @Override
    public void onPurchaseCancelled() {
        mUpgradeLog.info("Setting billing process finished...");
        bBillingProcessFinished = true;
    }

    @Override
    public void onPurchaseSuccessful(@Nullable List<Purchase> purchases) {
        if (purchases != null) {
            Purchase purchase = purchases.get(0);
            if (selectedProductDetails != null && selectedProductDetails.getOneTimePurchaseOfferDetails() != null) {
                mGoogleBillingManager.InAppConsume(purchase);
            } else {
                mGoogleBillingManager.subscriptionConsume(purchase);
            }
        }
    }

    @Override
    public void onRestorePurchaseClick() {
        mUpgradePresenter.restorePurchase();
    }

    @Override
    public void onTenGbFreeClick() {
        mUpgradeLog.info("User clicked on continue free...");
        mUpgradePresenter.onContinueFreeClick();
    }

    @Override
    public void onPolicyClick() {
        openURLInBrowser(NetworkKeyConstants.getWebsiteLink(NetworkKeyConstants.URL_PRIVACY));
    }

    @Override
    public void onTermsClick() {
        openURLInBrowser(NetworkKeyConstants.getWebsiteLink(NetworkKeyConstants.URL_TERMS));
    }

    @Override
    public void querySkuDetails(List<QueryProductDetailsParams.Product> products) {
        mUpgradeLog.info("Querying sku details...");
        mGoogleBillingManager.querySkuDetailsAsync(products);
    }

    @Override
    public void restorePurchase() {
        mAmazonBillingManager.getPurchaseHistory();
    }

    @Override
    public void setBillingProcessStatus(boolean bProcessFinished) {
        bBillingProcessFinished = bProcessFinished;
    }

    @Override
    public void setEmailStatus(boolean isEmailAdded, boolean isEmailConfirmed) {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.cl_upgrade);
        if (fragment instanceof PlansFragment) {
            ((PlansFragment) fragment).setEmailStatus(isEmailAdded, isEmailConfirmed);
        }
    }

    @Override
    public void showBillingDialog(final WindscribeInAppProduct windscribeInAppProduct, final boolean isEmailAdded,
            final boolean isEmailConfirmed) {
        PlansFragment.newInstance()
                .add(this, windscribeInAppProduct, R.id.cl_upgrade, false, isEmailAdded, isEmailConfirmed);
    }

    @Override
    public void showBillingErrorDialog(String errorMessage) {
        hideProgressBar();
        ErrorFragment.getInstance().add(errorMessage, this, R.id.cl_upgrade, false);
    }

    @Override
    public void showProgressBar(final String progressHeaderText) {
        runOnUiThread(() -> {
            Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.cl_upgrade);
            if (!(fragment instanceof ProgressFragment)) {
                ProgressFragment.getInstance().add(progressHeaderText, UpgradeActivity.this, R.id.cl_upgrade, false);
                getSupportFragmentManager().executePendingTransactions();
            }
            if (fragment instanceof ProgressFragment) {
                ((ProgressFragment) fragment).updateProgressStatus(progressHeaderText);
            }
        });

    }

    @Override
    public void showToast(String toastText) {
        mUpgradeLog.info("Showing toast to the user...");
        Toast.makeText(this, toastText, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void startPurchaseFlow(ImmutableList<BillingFlowParams.ProductDetailsParams> productDetailsParams, String accountID) {
        BillingFlowParams.Builder builder = BillingFlowParams.newBuilder()
                .setProductDetailsParamsList(productDetailsParams);
        if (accountID != null) {
            builder.setObfuscatedAccountId(accountID);
        }
        mUpgradeLog.info("Launching billing flow...");
        mUpgradePresenter.setPurchaseFlowState(PurchaseState.IN_PROCESS);
        mGoogleBillingManager.launchBillingFlow(this, builder.build());
    }

    @Override
    public void startPurchaseFlow(final Product product) {
        mAmazonBillingManager.launchPurchaseFlow(product);
    }

    @Override
    public void startSignUpActivity() {
        Intent startIntent = WelcomeActivity.getStartIntent(this);
        startIntent.putExtra("startFragmentName", "AccountSetUp");
        startActivity(startIntent);
        finish();
    }

    @Override
    public void startWindscribeActivity() {
        Intent startIntent = new Intent(this, WindscribeActivity.class);
        startIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | FLAG_ACTIVITY_NEW_TASK);
        startActivity(startIntent);
        finish();
    }

    void initAmazonBillingLifecycleListeners() {
        mAmazonBillingManager.onBillingSetUpSuccess.observe(this, code -> {
            mUpgradeLog.info("Billing client connected successfully...");
            mUpgradePresenter.onBillingSetupSuccessful();
        });

        mAmazonBillingManager.onProductsResponseSuccess
                .observe(this, products -> mUpgradePresenter.onProductDataResponse(products));

        mAmazonBillingManager.onProductsResponseFailure
                .observe(this, requestStatus -> mUpgradePresenter.onProductResponseFailure());

        mAmazonBillingManager.onPurchaseResponseSuccess
                .observe(this, purchaseResponse -> mUpgradePresenter.onPurchaseResponse(purchaseResponse));

        mAmazonBillingManager.onPurchaseResponseFailure
                .observe(this, requestStatus -> mUpgradePresenter.onPurchaseResponseFailure(requestStatus));

        mAmazonBillingManager.onAmazonPurchaseHistorySuccess
                .observe(this, purchases -> mUpgradePresenter.onAmazonPurchaseHistorySuccess(purchases));

        mAmazonBillingManager.onAmazonPurchaseHistoryError
                .observe(this, error -> mUpgradePresenter.onAmazonPurchaseHistoryError(error));
    }

    void initBillingLifecycleListeners() {
        mGoogleBillingManager.onBillingSetUpSuccess.observe(this, code -> {
            mUpgradeLog.info("Billing client connected successfully...");
            mUpgradePresenter.onBillingSetupSuccessful();
        });
        mGoogleBillingManager.onBillingSetupFailure.observe(this, code -> {
            mUpgradeLog.info("Billing client set up failure...");
            mUpgradePresenter.onBillingSetupFailed(code);
        });

        mGoogleBillingManager.onProductConsumeSuccess.observe(this, purchase -> {
            mUpgradeLog.info("Product consumption successful...");
            showToast(getResources().getString(R.string.purchase_successful));
            mUpgradePresenter.onPurchaseConsumed(purchase);
        });

        mGoogleBillingManager.onProductConsumeFailure.observe(this, customPurchase -> {
            mUpgradeLog.debug("Product consumption failed...");
            mUpgradePresenter.onConsumeFailed(customPurchase.getResponseCode(), customPurchase.getPurchase());
        });

        mGoogleBillingManager.purchaseUpdateEvent.observe(this, customPurchases -> {
            mUpgradeLog.info("Purchase updated...");
            mUpgradePresenter.onPurchaseUpdated(customPurchases.getResponseCode(), customPurchases.getPurchase());
        });
        mGoogleBillingManager.querySkuDetailEvent.observe(this, customSkuDetails -> mUpgradePresenter
                .onSkuDetailsReceived(customSkuDetails.getBillingResult().getResponseCode(),
                        customSkuDetails.getProductDetails()));

    }

    private void setBillingType() {
        PackageManager pkgManager = getPackageManager();
        String installerPackageName = pkgManager.getInstallerPackageName(getPackageName());
        if (installerPackageName != null && installerPackageName.startsWith("com.amazon")) {
            billingType = BillingType.Amazon;
        } else {
            billingType = BillingType.Google;
        }
    }
}

File: tv/src/main/java/com/windscribe/tv/upgrade/UpgradeActivity.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.tv.upgrade

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import com.amazon.device.iap.model.Product
import com.amazon.device.iap.model.PurchaseResponse
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingFlowParams.ProductDetailsParams
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.QueryProductDetailsParams
import com.google.common.collect.ImmutableList
import com.windscribe.tv.R
import com.windscribe.tv.base.BaseActivity
import com.windscribe.tv.confirmemail.ConfirmActivity
import com.windscribe.tv.customview.CustomDialog
import com.windscribe.tv.customview.ErrorFragment
import com.windscribe.tv.customview.ProgressFragment
import com.windscribe.tv.di.ActivityModule
import com.windscribe.tv.email.AddEmailActivity
import com.windscribe.tv.upgrade.PlansFragment.Companion.newInstance
import com.windscribe.tv.welcome.WelcomeActivity
import com.windscribe.tv.windscribe.WindscribeActivity
import com.windscribe.vpn.Windscribe.Companion.appContext
import com.windscribe.vpn.api.response.PushNotificationAction
import com.windscribe.vpn.billing.*
import com.windscribe.vpn.constants.ExtraConstants.PROMO_EXTRA
import okhttp3.internal.toImmutableList
import org.slf4j.LoggerFactory
import javax.inject.Inject

class UpgradeActivity : BaseActivity(), UpgradeView, BillingFragmentCallback {
    enum class BillingType(val value: String) {
        Google("android"), Amazon("amazon")
    }
    @Inject
    lateinit var upgradeDialog: CustomDialog
    @Inject
    lateinit var presenter: UpgradePresenter
    override var isBillingProcessFinished = false
        private set
    override var billingType = BillingType.Google
        private set
    private var amazonBillingManager: AmazonBillingManager? = null
    private var googleBillingManager: GoogleBillingManager? = null
    private val logger = LoggerFactory.getLogger("upgrade_a")
    private var selectedProductDetails: ProductDetails? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setActivityModule(ActivityModule(this, this)).inject(this)
        setContentLayout(R.layout.activity_upgrade)
        logger.info("OnCreate: Upgrade Activity")
        showProgressBar("Loading Billing Plans...")
        if (intent.hasExtra(PROMO_EXTRA)) {
            val pushNotificationAction = intent
                .getSerializableExtra(PROMO_EXTRA) as PushNotificationAction?
            pushNotificationAction?.let {
                presenter.setPushNotificationAction(it)
            }
        }
        setBillingType()
        if (billingType == BillingType.Amazon) {
            amazonBillingManager = appContext.amazonBillingManager
            amazonBillingManager?.let {
                lifecycle.addObserver(it)
                initAmazonBillingLifecycleListeners()
            }
        } else {
            googleBillingManager = appContext.billingManager
            googleBillingManager?.let {
                lifecycle.addObserver(it)
                initBillingLifecycleListeners()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (billingType == BillingType.Google) {
            presenter.checkBillingProcessStatus()
        }
        presenter.setLayoutFromApiSession()
    }

    override fun onDestroy() {
        upgradeDialog.cancel()
        presenter.onDestroy()
        super.onDestroy()
    }

    override fun getProducts(skuList: List<String>) {
        amazonBillingManager?.getProducts(skuList)
    }

    override fun goBackToMainActivity() {
        logger.info("Going back to previous activity...")
        onBackPressed()
    }

    override fun goToAddEmail() {
        val startIntent = Intent(this, AddEmailActivity::class.java)
        startIntent.putExtra("goToHomeAfterFinish", false)
        startActivity(startIntent)
    }

    override fun goToConfirmEmail() {
        val startIntent = Intent(this, ConfirmActivity::class.java)
        startActivity(startIntent)
    }

    override fun gotToClaimAccount() {
        val startIntent = WelcomeActivity.getStartIntent(this)
        startIntent.putExtra("startFragmentName", "AccountSetUp")
        startActivity(startIntent)
        finish()
    }

    override fun hideProgressBar() {
        val fragment = supportFragmentManager.findFragmentById(R.id.cl_upgrade)
        if (fragment is ProgressFragment) {
            supportFragmentManager.popBackStack()
        }
    }

    override fun onContinuePlanClick(productDetails: ProductDetails, selectedIndex: Int) {
        logger.info("User clicked on plan item...")
        selectedProductDetails = productDetails
        val builder = ProductDetailsParams.newBuilder()
        builder.setProductDetails(productDetails)
        productDetails.subscriptionOfferDetails?.let {
            val offerToken = it[selectedIndex].offerToken
            builder.setOfferToken(offerToken)
        }
        val productDetailsParamsList = ImmutableList.of(builder.build())
        presenter.onMonthlyItemClicked(productDetailsParamsList)
    }

    override fun onContinuePlanClick(selectedSku: Product) {
        presenter.onContinuePlanClick(selectedSku)
    }

    override fun onPurchaseCancelled() {
        logger.info("Setting billing process finished...")
        isBillingProcessFinished = true
    }

    override fun onPurchaseSuccessful(purchases: List<Purchase>) {
        val purchase = purchases[0]
        if (selectedProductDetails?.oneTimePurchaseOfferDetails != null) {
            googleBillingManager?.InAppConsume(purchase)
        } else {
            googleBillingManager?.subscriptionConsume(purchase)
        }
    }


    override fun onRestorePurchaseClick() {
        presenter.restorePurchase()
    }

    override fun onTenGbFreeClick() {
        logger.info("User clicked on continue free...")
        presenter.onContinueFreeClick()
    }

    override fun onPolicyClick() {}

    override fun onTermsClick() {}

    override fun querySkuDetails(products: List<QueryProductDetailsParams.Product>) {
        logger.info("Querying sku details...")
        googleBillingManager?.querySkuDetailsAsync(products)
    }

    override fun restorePurchase() {
        amazonBillingManager?.getPurchaseHistory()
    }

    override fun setBillingProcessStatus(processFinished: Boolean) {
        isBillingProcessFinished = processFinished
    }

    override fun setEmailStatus(isEmailAdded: Boolean, isEmailConfirmed: Boolean) {}

    override fun showBillingDialog(
        windscribeInAppProduct: WindscribeInAppProduct,
        isEmailAdded: Boolean,
        isEmailConfirmed: Boolean
    ) {
        newInstance()
            .add(
                this,
                windscribeInAppProduct,
                R.id.cl_upgrade,
                false,
                isEmailAdded,
                isEmailConfirmed
            )
    }

    override fun showBillingErrorDialog(errorMessage: String) {
        hideProgressBar()
        ErrorFragment.instance.add(errorMessage, this, R.id.cl_upgrade, false)
    }

    override fun showProgressBar(message: String) {
        runOnUiThread {
            val fragment = supportFragmentManager.findFragmentById(R.id.cl_upgrade)
            if (fragment !is ProgressFragment) {
                ProgressFragment.instance.add(
                    message,
                    this@UpgradeActivity,
                    R.id.cl_upgrade,
                    false
                )
                supportFragmentManager.executePendingTransactions()
            }
            if (fragment is ProgressFragment) {
                fragment.updateProgressStatus(message)
            }
        }
    }

    override fun showToast(toastText: String) {
        logger.info("Showing toast to the user...")
        Toast.makeText(this, toastText, Toast.LENGTH_SHORT).show()
    }

    override fun startPurchaseFlow(
        productDetailsParams: ImmutableList<ProductDetailsParams>,
        accountID: String?
    ) {
        val builder =
            BillingFlowParams.newBuilder().setProductDetailsParamsList(productDetailsParams)
        accountID?.let { builder.setObfuscatedAccountId(it) }
        logger.info("Launching billing flow...")
        presenter.setPurchaseFlowState(PurchaseState.IN_PROCESS)
        googleBillingManager?.launchBillingFlow(this, builder.build())
    }

    override fun startPurchaseFlow(product: Product) {
        amazonBillingManager?.launchPurchaseFlow(product)
    }

    override fun startSignUpActivity() {
        val startIntent = WelcomeActivity.getStartIntent(this)
        startIntent.putExtra("startFragmentName", "AccountSetUp")
        startActivity(startIntent)
        finish()
    }

    override fun startWindscribeActivity() {
        val startIntent = Intent(this, WindscribeActivity::class.java)
        startIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(startIntent)
        finish()
    }

    private fun initAmazonBillingLifecycleListeners() {
        amazonBillingManager?.onBillingSetUpSuccess?.observe(this) {
            logger.info("Billing client connected successfully...")
            presenter.onBillingSetupSuccessful()
        }
        amazonBillingManager?.onProductsResponseSuccess?.observe(
            this
        ) { products: Map<String, Product> ->
            presenter.onProductDataResponse(products)
        }
        amazonBillingManager?.onProductsResponseFailure?.observe(
            this
        ) { presenter.onProductResponseFailure() }
        amazonBillingManager?.onPurchaseResponseSuccess?.observe(this) { purchaseResponse: PurchaseResponse ->
            presenter.onPurchaseResponse(
                    purchaseResponse
            )
        }
        amazonBillingManager?.onPurchaseResponseFailure?.observe(this) { requestStatus: PurchaseResponse.RequestStatus ->
            presenter.onPurchaseResponseFailure(
                    requestStatus
            )
        }
        amazonBillingManager?.onAmazonPurchaseHistorySuccess?.observe(
            this
        ) { purchases: List<AmazonPurchase> ->
            presenter.onAmazonPurchaseHistorySuccess(purchases)
        }
        amazonBillingManager?.onAmazonPurchaseHistoryError?.observe(this) { error: String ->
            presenter.onAmazonPurchaseHistoryError(
                    error
            )
        }
    }

    private fun initBillingLifecycleListeners() {
        googleBillingManager?.onBillingSetUpSuccess?.observe(this) {
            logger.info("Billing client connected successfully...")
            presenter.onBillingSetupSuccessful()
        }
        googleBillingManager?.onBillingSetupFailure?.observe(this) { code: Int ->
            logger.info("Billing client set up failure...")
            presenter.onBillingSetupFailed(code)
        }
        googleBillingManager?.onProductConsumeSuccess?.observe(this) { purchase: Purchase ->
            logger.info("Product consumption successful...")
            showToast(resources.getString(R.string.purchase_successful))
            presenter.onPurchaseConsumed(purchase)
        }
        googleBillingManager?.onProductConsumeFailure?.observe(
            this
        ) { customPurchase: CustomPurchase ->
            logger.debug("Product consumption failed...")
            presenter.onConsumeFailed(
                    customPurchase.responseCode,
                    customPurchase.purchase
            )
        }
        googleBillingManager?.purchaseUpdateEvent?.observe(
            this
        ) { customPurchases: CustomPurchases ->
            logger.info("Purchase updated...")
            presenter.onPurchaseUpdated(
                    customPurchases.responseCode,
                    customPurchases.purchase
            )
        }
        googleBillingManager?.querySkuDetailEvent?.observe(
            this
        ) { customProductDetails: CustomProductDetails ->
            presenter
                .onSkuDetailsReceived(
                    customProductDetails.billingResult.responseCode,
                    customProductDetails.productDetails.toImmutableList()
                )
        }
    }

    private fun setBillingType() {
        val pkgManager = packageManager
        val installerPackageName = pkgManager.getInstallerPackageName(packageName)
        billingType =
            if (installerPackageName != null && installerPackageName.startsWith("com.amazon")) {
                BillingType.Amazon
            } else {
                BillingType.Google
            }
    }

    companion object {
        @JvmStatic
        fun getStartIntent(context: Context): Intent {
            return Intent(context, UpgradeActivity::class.java)
        }
    }
}
